### DISCUSSION TITLE
Release 3.5.0.-beta: hierarchical import through OAI and CUSTOM interfaces

### DISCUSSION THREAD

**Discussion by aetherfaerber** [kitodo/kitodo-production#5508] [Q&A] [answered]
We have been experimenting with a custom import interface but more recently switched to an OAI interface. We have been trying to make multi-level imports work with both, but found several obstacles. In case of OAI, these may also be grounded in the standard.

- the import configuration setup page for an OAI interface has neither a field to configure the superordinate ID parameter („ID-Parameter der Überordnung“) nor search fields.
- the import configuration setup page for a CUSTOM interface has a field to configure the superodinate ID parameter but no new search fields may be added, so there is nothing to choose. (Does this imply that the superordinate field should be identical to the subordinate, e.g. “itemID”?)
- We have defined a metadata field as higherlevelIdentifier in the ruleset.xml but choosing import depth 2 during import results in two identical queries for the subordinate ID. What are we missing?
- the parent element type in the OAI interface can only be set to `<reference>`. What kind of element is this and how can we (and should we?) implement this?

**Comment by solth** [comment]
>     * the import configuration setup page for an OAI interface has neither a field to configure the superordinate ID parameter („ID-Parameter der Überordnung“) nor search fields.

This limitation exists because OAI PMH interfaces are not really meant for searching but for harvesting instead and thus do not allow defining custom search fields (including fields querying records by parent ID). To support metadata import from OAI interfaces anyway we make use of the predefined OAI "verb" `GetRecord` to retrieve individual records, which can not be modified. (see http://www.openarchives.org/OAI/openarchivesprotocol.html#GetRecord for details). If your OAI interface has different search fields defined it may be an indication that it is not a real OAI interface that adheres to the OAI specifications (see comments of #5271, too)

>     * the import configuration setup page for a CUSTOM interface has a field to configure the superodinate ID parameter but no new search fields may be added, so there is nothing to choose. (Does this imply that the superordinate field should be identical to the subordinate, e.g. “itemID”?)

This is a bug. (and `ID parameter` and `ID parameter for parent record` are not supposed to be the same search field.)
The interface type `CUSTOM` was mainly introduced to enable querying a "fake" OAI interface, see #5257 and the linked PR for details. 

Because being able to search for individual records by ID was sufficient for the use case at that time, `CUSTOM` search interfaces did not allow automatic retrieval of parent records (for which `ID parameter for parent record` is used) or hit lists (triggered by non-ID search fields).  The `+` button under the search field list which is deactivated after one field has been added, should have a corresponding explanation in it's mouse hover tooltip. 

Since enabling these features for `CUSTOM` type search interfaces would probably require quite a few changes in the code, the best short time solution for this would be to remove the `ID parameter for parent record` configuration from these search interfaces until more comprehensive changes can be made to add these features, IMHO.

>     * We have defined a metadata field as higherlevelIdentifier in the ruleset.xml but choosing import depth 2 during import results in two identical queries for the subordinate ID. What are we missing?

This is difficult to say without further investigation. There were several bugs concerning import configurations in release 3.5.0-beta which have since been fixed (see https://github.com/kitodo/kitodo-production/milestone/18 for progress). Perhaps you could try with the most current master branch to see if the problem you described has been resolved as well?

>     * the parent element type in the OAI interface can only be set to `<reference>`. What kind of element is this and how can we (and should we?) implement this?

This is probably a remnat of earlier version of Kitodo.Production 3 and may not be required anymore. (It should not matter if you select 'reference' here). The corresponding GitHub issue is #5318. I will have to check whether it is still required in some cases when I find the time.

**Reply by aetherfaerber** [reply]
First of all thank you very very much for pointing me to the relevant issues and the OAI spec. This effort is really appreciated and I will make sure to search and consult the specs next time beforehand.

Alas after going through the spec and the linked issues I don't see my issue resolved but rather manifested in current limitations of the interfaces and hence am back with more questions.

The most important one is ”does this mean hierarchical import is only possible with SRU interfaces but not with OAI or CUSTOM?”

And more explicitely:

> This limitation exists because OAI PMH interfaces are not really meant for searching but for harvesting instead and thus do not allow defining custom search fields (including fields querying records by parent ID). To support metadata import from OAI interfaces anyway we make use of the predefined OAI "verb" GetRecord to retrieve individual records, which can not be modified. (see http://www.openarchives.org/OAI/openarchivesprotocol.html#GetRecord for details). If your OAI interface has different search fields defined it may be an indication that it is not a real OAI interface that adheres to the OAI specifications (see comments of https://github.com/kitodo/kitodo-production/issues/5271, too)

Thank you. After reading the spec I fully agree that OAI interfaces should not have any custom search fields. But we really don't need/want to search (might look like it when used to SRU, though). 
Everything we need is described in the spec as “selective harvesting“, namely [selective harvesting by set](https://www.openarchives.org/OAI/openarchivesprotocol.html#SelectiveHarvestingandSets). *But* this is only available through the `ListItems` and `ListRecords` verbs and not the currently implemented `GetRecord`. Is inclusion of these verbs planned?

Currently, as it seems, the best we could do is to add `ListItems`or `ListRecords` as possible parameter values for parameter key `verb` as URL parameters for a CUSTOM interface. 

But as long as
> […] CUSTOM search interfaces did not allow automatic retrieval of parent records (for which ID parameter for parent record is used) or hit lists (triggered by non-ID search fields). 

stays valid I guess… we are out of luck?

> Perhaps you could try with the most current master branch to see if the problem you described has been resolved as well?

We switched to the most current master branch but still get the same result for our OAI interface. Although, with the information provided above I am not even sure if this should work at all.

**Comment by solth** [comment, marked as answer]
Yes, adding support for the OAI verb `ListRecords` has been on my mind for some time. It could be used to create hit lists similar to those resulting from regular search queries, instead showing all records of the selected OAI set. 

This would probably not be a trivial change, though, so implementing this feature will most likely require some time and resources. Perhaps this could be a candidate for the development fund 2023.

Could you provide me with an example URL of the search interface you want to query for an individual record? I could check whether allowing custom search fields for CUSTOM type interfaces is more feasible in the current master or not in order to create the query you need.

**Reply by aetherfaerber** [reply]
> Yes, adding support for the OAI verb `ListRecords` has been on my mind for some time. It could be used to create hit lists similar to those resulting from regular search queries, instead showing all records of the selected OAI set.
> 
> This would probably not be a trivial change, though, so implementing this feature will most likely require some time and resources. Perhaps this could be a candidate for the development fund 2023.

I already guessed it wouldn't be trivial and will happily submit a proposal for this years fund.

> Could you provide me with an example URL of the search interface you want to query for an individual record? I could check whether allowing custom search fields for CUSTOM type interfaces is more feasible in the current master or not in order to create the query you need.

Currently we use the BaseX HTTP-Interface which allows configurable queries via [RESTXQ](https://docs.basex.org/wiki/RESTXQ):
    
    http://xyz.uni-marburg.de:8080/search?identifier=b1409
    http://xyz.uni-marburg.de:8080/search?identifier=v4637623
    
I have placed the resulting XML files here:
    https://digitalisate-he.arcinsys.de/download/b1409.xml
    https://digitalisate-he.arcinsys.de/download/v4637623.xml
    
As you can see, both files include the subordinate ID (v4637623) and the superordinate ID (b1409). We would like to create processes with the same hierarchy. Ideally, importing b1409 would allow to create processes for all its subordinate IDs as listed in the XML.

**Reply by solth** [reply]
> http://xyz.uni-marburg.de:8080/search?identifier=b1409
> http://xyz.uni-marburg.de:8080/search?identifier=v4637623

Creating those query URLs for individual data records should already be possible with the current functionality of CUSTOM search interfaces. I did encounter a bug when no custom URL parameters have been configured in a CUSTOM type search interface, though, which should be fixed in the branch https://github.com/effective-webwork/kitodo-production/tree/custom-url-fix. When using this branch as a base, I was able to create exactly the URLs you provided. Perhaps you could try out this potential fix and check whether you can create correct URLs for your search interface with it (the created query urls should appear in your kitodo.log file when you change the log level to "debug")

Concerning your second question about creating multiple processes from one xml file (e.g. one process for each `<c level="file">`) this is not possible with Kitodo.Production at the moment. At most you can create two processes - i.e. one for the record identified by the provided ID itself (`level="file"` in your EAD case) and one for a potentially contained parent record (`level="collection"` or `level="class"`, depending on which structures you want to map) - from one xml file by providing a "parent mapping file" in the corresponding "Import Configuration".

In case you can adjust the EAD xml response returned by your interface in such a way that the returned xml file (for example `https://digitalisate-he.arcinsys.de/download/v4637623.xml`) does not contain 620 `file` elements, but instead just the one with the specified ID (`<c level="file" id="v4637623">`), I would recommend using the "Mass import" of Kitodo.Production 3, though. This feature allows you to import multiple records specified by each records ID in a CSV file into indiviual processes automatically. If the second line in the CSV file contains the ID of the `collection` (the first line is interpreted as column headers), and the following lines contain the IDs of the individual `file`s, hierarchical relations between imported records will also be automatically mapped to corresponding Kitodo process hierarchies during import. 
It seems to me, this way would be the best solution for your use case.

**Reply by aetherfaerber** [reply]
> Creating those query URLs for individual data records should already be possible with the current functionality of CUSTOM search interfaces. I did encounter a bug when no custom URL parameters have been configured in a CUSTOM type search interface, though, which should be fixed in the branch https://github.com/effective-webwork/kitodo-production/tree/custom-url-fix. When using this branch as a base, I was able to create exactly the URLs you provided. Perhaps you could try out this potential fix and check whether you can create correct URLs for your search interface with it (the created query urls should appear in your kitodo.log file when you change the log level to "debug")

Thank you for linking to this branch. We didn't run into any difficulties concerning superfluous "&"s so far but managed to filter out an additional "=" at the interfaces'  side.
We were also able to construct the correct URLs and import of individual items on vID-level does work (although imports of the parent item cannot be saved - probably because of the contained IDs).


> Concerning your second question about creating multiple processes from one xml file (e.g. one process for each `<c level="file">`) this is not possible with Kitodo.Production at the moment. At most you can create two processes - i.e. one for the record identified by the provided ID itself (`level="file"` in your EAD case) and one for a potentially contained parent record (`level="collection"` or `level="class"`, depending on which structures you want to map) - from one xml file by providing a "parent mapping file" in the corresponding "Import Configuration".

Thank you for clarifying this! So there is no way to automatically have a parent import all children but a child can import one parent (item, maybe several levels up but only one item per level).
What then is the use of „Kindvorgänge importieren“?

> In case you can adjust the EAD xml response returned by your interface in such a way that the returned xml file (for example `https://digitalisate-he.arcinsys.de/download/v4637623.xml`) does not contain 620 `file` elements, but instead just the one with the specified ID (`<c level="file" id="v4637623">`), I would recommend using the "Mass import" of Kitodo.Production 3, though. This feature allows you to import multiple records specified by each records ID in a CSV file into indiviual processes automatically. If the second line in the CSV file contains the ID of the `collection` (the first line is interpreted as column headers), and the following lines contain the IDs of the individual `file`s, hierarchical relations between imported records will also be automatically mapped to corresponding Kitodo process hierarchies during import. It seems to me, this way would be the best solution for your use case.

Thanks a lot for pointing us in this direction. I had considered this as last resort but wanted to try all other possible options. We adjusted the xml output of our interface as described and can now import both file-level items (vID) and collection level items (bID) just fine. We can also create subordinate processes for collection level processes via the gui, regardless if they are actually connected or not. Importing a file-level item with import depth 2 still results in two identical imports of just the file-level item.
We are going to test the mass import feature but it seems broken (NullPointer Exception) in the beta release. So we will have to make sure to switch to the branch you mentioned above first.

**Reply by BartChris** [reply]
> Thank you for clarifying this! So there is no way to automatically have a parent import all children but a child can import one parent (item, maybe several levels up but only one item per level). What then is the use of „Kindvorgänge importieren“?

As far as i can see this relies on the `parentIDSearchField` which is not avaible for custom import configurations.
https://github.com/kitodo/kitodo-production/blob/37bfd849fea8f174c9261e0c978567c89b0e6721/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L592

**Reply by solth** [reply]
> As far as i can see this relies on the `parentIDSearchField` which is not avaible for custom import configurations.

Yes, exactly. Whether child processes can be imported automatically depends on whether a search interface can be queried with the ID of a superordinate record and then return the xml records of all subordinate records. The only example I know of is the SRU interfaces of Kalliope, though. There the URL parameter `context.ead.id` can be used to specifify the record whose child records are returned. 

So for example the following URL will return a list of records that are subordinate to the process with the id `DE-2060-BE-13-C35`:
https://kalliope-verbund.info/sru?recordSchema=mods&operation=searchRetrieve&version=1.2&query=context.ead.id%3DDE-2060-BE-13-C35

When such a URL parameter exists in a search interface, it has to be configured as a search field for an `ImportConfiguration` in Kitodo and this search field than has to be set as "ID parameter of parent record", which in turn represents the database column `parentIDSearchField` of a configuration.

**Reply by BartChris** [reply]
>  We are going to test the mass import feature but it seems broken (NullPointer Exception) in the beta release. So we will have to make sure to switch to the branch you mentioned above first.

Can you elaborate on that? What have you tried and where exactly does the NPE happen?

**Reply by solth** [reply]
I would like to know, too. We did have a few problems related to the UI table holding the IDs of records to import and the curresponding `blur` javascripts but those should be fixed in the current master.

**Reply by aetherfaerber** [reply]
> > We are going to test the mass import feature but it seems broken (NullPointer Exception) in the beta release. So we will have to make sure to switch to the branch you mentioned above first.
> 
> Can you elaborate on that? What have you tried and where exactly does the NPE happen?

First of all sorry the very belated reply. I was mostly unable to work in February.

Regarding the question about NPEs sadly I cannot reproduce this behaviour now, although we are quite sure that nothing has changed in our setup since I reported it.
I tried really hard to get a NPE yesterday and succeeded once after going directly to massimport.jsf and starting imports without a production template but I cannot reproduce this now either and am quite sure this is not what I did a month ago.

**Reply by aetherfaerber** [reply]
I have marked this one as answer since it contains the pointer towards the mass-import feature, which we finally managed to use for hierarchical imports.

Also, i will make sure to submit adding support for `ListRecords` as a candidate for the development fund.

**Reply by solth** [reply]
> We are going to test the mass import feature but it seems broken (NullPointer Exception) in the beta release. So we will have to make sure to switch to the branch you mentioned above first.

I was unable to reproduce a `NullPointerException` on the mass import page with the current state of the code, but there does seem to be one instance of unhandled exception in form of a `ConfigException` (documented in #5584). Could this maybe the same error you encountered?

**Reply by aetherfaerber** [reply]
> Also, i will make sure to submit adding support for `ListRecords` as a candidate for the development fund.

I have opened an issue: https://github.com/kitodo/kitodo-production/issues/5587
@solth could you please add the Development fund label to the issue?

**Reply by solth** [reply]
> @solth could you please add the Development fund label to the issue?

done!

**Comment by BartChris** [comment]
>>    […] CUSTOM search interfaces did not allow automatic retrieval of parent records (for which ID parameter for parent record is used) >> or > hit lists (triggered by non-ID search fields).

> stays valid I guess… we are out of luck?

In addition to everything said above: It appears to me that in your case the same field ("identifier") is used for retrieving children and parent records. So the automatic retrieval of a parent record should also work (at least for the normal process creation). I am also using a custom interface and when i import a child record the parent record is also imported. So this works for me. 

Edit: But i am also not sure WHY this works - i am still not fully sure about the exact purpose of the functional keys `higherlevelidentifier`, `recordIdentifer` and the `ID parameter for parent record` in the context discuessed here. For retrieving a parent record which is already in the database `higherlevelidentifier` and `recordIdentifer` are used. But i am not sure right what exactly is used for creating a new parent record from an imported child. @solth knows best about this, so he can probably answer that.

**Reply by aetherfaerber** [reply]
Thank you very much for the additional information @BartChris . 

Regarding a clarification of the functional keys, the only documentation I know of is a comment in the ruleset file. It at least marks the purpose of higherlevelIdentifier:

```
Description of the applied attributes:
1. processTitle="TSL_ATS+'_'+CatalogIDDigital" -> Definition of the process title.
2. use="createChildrenFromParent" -> For document types like periodical, when children proce    sses should be created by the parent process in the process list.
3. use="createChildrenWithCalendar" -> For document types like newspaper, when the calendar     should be available for parent processes in the process list to create processes for issues.
4. use="higherlevelIdentifier" -> For metadata which contains the identifier of the superordinate process.
5. use="processTitle" -> For document types which are created with the calender without Cata    logIDDigital, like newspaper issues. The process title can be transformed in the export XSL.
6. withWorkflow="false" -> For superordinate document type like periodical, newspaper, multivolume works, whose processes do not need a workflow.
```

This does help with assigning the keys but doesn't explain anything about the inner workings and interrelations. As you said, a further explanation would be appreciated.


The automatic import of a parent record (which isn't our use case but could be if we can mass-import all children and they get automatically assigned to only one singe parent process) didn't work for us last time i checked, but I will try this again in order to report back.

**Comment by BartChris** [comment]
My understanding of those keys is that `higherlevelidentifier` and `recordIdentifer` are attributes which you assign to keys defined in your ruleset. The `higherlevelidentifier` marks the field which contains the key of the parent record of the current record. The `recordIdentifier` marks the key which contains the identifier of the digital object in the source metadata system. 

The `ID parameter for parent record` is not about keys in your ruleset but about search fields in an external search interface. You basically define the key which is used to search in all records of an external catalogue for those titles which have a link to a specific parent record. You need that key if you want to import all child records for a specific parent record from a catalog.

So when the system is searching if there is already an parent record in the database for the newly created record the following is happing:

1.  the ID inside the field marked as `higherlevelidentifier` is extracted; this is the identifier of the parent object we want to search in our Kitodo database/Kitodo search index
2. This ID is then searched for in all processes and there specifically in all fields which are marked as `recordIdentifier` fields (which could be different fields for different document types):

https://github.com/effective-webwork/kitodo-production/blob/6c8240351c764e5838a7bc66afa58d0399615b7f/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L934-L960

When the parent's record identifier is **not** already contained in the database, Kitodo is trying to import the parent from an external catalog. The number of times this import logic is called depends on the number of import levels defined during import. This is done using repeated/recursive calls of the method `importProcessAndReturnParentID` with supplying the current `parentid` as `recordid` which now should be imported,

The logic works in a way that after the import of a record, the `higherlevelidentifier` of this record is extracted using the defined value for `higherlevelidentifier` and then saved as local variable `parentid`

https://github.com/effective-webwork/kitodo-production/blob/6c8240351c764e5838a7bc66afa58d0399615b7f/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L376-L392

This `parentid` can then be used to query the search interface for a parent record with this ID reusing the normal import logic. This implies in my understanding that the parent record and the child record need to use the same field in the metadata source to store their respective identifier, which is then queried.

https://github.com/effective-webwork/kitodo-production/blob/6c8240351c764e5838a7bc66afa58d0399615b7f/Kitodo/src/main/java/org/kitodo/production/services/data/ImportService.java#L544

**Reply by aetherfaerber** [reply]
Thank you for these explanations. They really gave us a better understanding of what Kitodo is doing here. Also your explanation of `recordIdentifier` was what we needed to finally get mass imports working.
